sdk/helper/storagepacker/invalidate.go

```diff
@@ -0,0 +1,295 @@
+
+	originalBucket, err := s.GetBucket(ctx, bucketKey)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem finding original bucket: {{err}}", err)
+	}
+	originalBucket.Lock()
+	defer originalBucket.Unlock()
+
+	// Any item included in the new bucket should be reported present.
+	if !replacementBucket.HasShards && replacementBucket.ItemMap != nil {
+		for k, v := range replacementBucket.ItemMap {
+			present = append(present, &Item{
+				ID:   k,
+				Data: v,
+			})
+		}
+	}
+
+	// Is the new bucket a replacement, or new?  If it's new, nothing to compare against.
+	if originalBucket.Key == replacementBucket.Key {
+		// Any item not present in the new bucket *might* be deleted, or
+		// it might be elsewhere (in which case we want to report that value instead,
+		// to make sure the correct version is being used.)
+		maybeDeleted := itemDifferenceBetween(originalBucket, replacementBucket)
+		var revisit []*Item
+		deleted, revisit, err = s.identifyItemsAbsentOrShadowed(bucketKey, maybeDeleted)
+		if err != nil {
+			// Return partial information?
+			return present, nil, err
+		}
+		present = append(present, revisit...)
+	}
+
+	// Swap the replacement bucket in to the cache (or delete)
+	cacheKey := s.GetCacheKey(bucketKey)
+	s.bucketsCacheLock.RLock()
+	if bucketRemoved {
+		s.bucketsCache.Delete(cacheKey)
+	} else {
+		s.bucketsCache.Insert(cacheKey, replacementBucket)
+	}
+	s.bucketsCacheLock.RUnlock()
+
+	return present, deleted, nil
+
+}
+
+// Look up each item and determine whether another copy (or version) of it exists
+// in a different bucket.
+func (s *StoragePackerV2) identifyItemsAbsentOrShadowed(bucketKey string, maybeDeleted []*itemRequest) ([]*Item, []*Item, error) {
+	deleted := make([]*Item, 0)
+	revisit := make([]*Item, 0)
+
+	// Using the radix lookup will find child buckets, but not parent
+	// buckets.  For that we'll have to walk up the tree a step at a time.
+	//
+	// For example, we might be processing 00/0
+	//
+	// 00 -- 00/0 -- 00/0/a
+	//
+	// but, the item with key 000a... might exist in 00/0/a or maybe even in 00,
+	// so we want to check both.
+	//
+	// If we find it somewhere, return the version in the bucket with the longest key.
+	// (That is more likely to be a more recent version.)
+	partition, err := s.partitionRequests(maybeDeleted)
+	if err != nil {
+		return nil, nil, errwrap.Wrapf("problem searching for deleted keys: {{err}}", err)
+	}
+
+	// Looking through the other buckets without acquiring a lock
+	// is unsafe, but should be OK in the context of a replica.
+	for _, p := range partition {
+		bucketsToCheck := s.bucketAndAllParents(p.Bucket)
+		for _, request := range p.Requests {
+			found := false
+			for _, b := range bucketsToCheck {
+				// Skip originalBucket
+				if b.Key != bucketKey {
+					var data []byte
+					if data, found = b.ItemMap[request.ID]; found {
+						revisit = append(revisit, &Item{
+							ID:   request.ID,
+							Data: data,
+						})
+						break
```

Comment:
Will this break the way you want? Are you trying to break from just the inner loop or continue looping in the next outer? Go's a little funny with its break statement; I think Python breaks from all loops.... _maybe_, I can't remember. Anyways, here's how this will behave: https://play.golang.org/p/UPcQrnE6N4q. If you're wanting to get to a different loop than the next outer one, you might consider using a Go label in conjunction with a break or GOTO statement. [This](https://medium.com/golangspec/labels-in-go-4ffd81932339) was the least awful reference I could find to it :-).

Reply:
I believe it's the same way in Python, break applies to just the innermost loop.  That's what I want here in any case, we want to stop searching buckets for a matching ID (Request object), but go on and do the search for the next ID.

Reply:
Perfect!